File: desktop/EatfitDesktop/Services/GrpcMealService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Eatfit.V1;
using Grpc.Core;
using Grpc.Net.Client;

namespace EatfitDesktop.Services
{
    public class GrpcMealService : IDisposable
    {
        private readonly GrpcChannel _channel;
        private readonly Eatfit.V1.MealService.MealServiceClient _client;
        private bool _disposed;

        public GrpcMealService(string serverUrl = "http://localhost:50051")
        {
            var handler = new HttpClientHandler
            {
                ServerCertificateCustomValidationCallback =
                    HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
            };

            _channel = GrpcChannel.ForAddress(serverUrl, new GrpcChannelOptions { HttpHandler = handler });
            _client = new Eatfit.V1.MealService.MealServiceClient(_channel);
        }

        public async Task<List<MealData>> ListMealsAsync(string userId, string sortBy = "created_at", string sortOrder = "desc")
        {
            try
            {
                var request = new ListMealsRequest
                {
                    UserId = userId,
                    SortBy = sortBy,
                    SortOrder = sortOrder,
                    Page = 1,
                    PageSize = 100
                };
                var response = await _client.ListMealsAsync(request);
                return new List<MealData>(response.Meals);
            }
            catch (RpcException)
            {
                return new List<MealData>();
            }
        }

        public async Task<List<MealData>> SearchMealsAsync(string userId, string query, string sortBy = "created_at", string sortOrder = "desc")
        {
            try
            {
                var request = new SearchMealsRequest
                {
                    UserId = userId,
                    Query = query,
                    SortBy = sortBy,
                    SortOrder = sortOrder,
                    Page = 1,
                    PageSize = 100
                };
                var response = await _client.SearchMealsAsync(request);
                return new List<MealData>(response.Meals);
            }
            catch
            {
                return new List<MealData>();
            }
        }

        public async Task<MealData?> GetMealAsync(string mealId)
        {
            try
            {
                var response = await _client.GetMealAsync(new GetMealRequest { MealId = mealId });
                return response.Meal;
            }
            catch
            {
                return null;
            }
        }

        public async Task<bool> CreateMealAsync(string userId, string name, string description, string recipe,
            string imageUrl, double calories, double proteins, double fats, double carbs, double water,
            List<MealComponentInput> components)
        {
            try
            {
                var request = new CreateMealRequest
                {
                    UserId = userId,
                    Name = name,
                    Description = description,
                    Recipe = recipe,
                    ImageUrl = imageUrl,
                    Calories = calories,
                    Proteins = proteins,
                    Fats = fats,
                    Carbs = carbs,
                    Water = water
                };
                request.Components.AddRange(components);
                await _client.CreateMealAsync(request);
                return true;
            }
            catch
            {
                return false;
            }
        }

        public async Task<bool> UpdateMealAsync(string mealId, string name, string description, string recipe,
            string imageUrl, double calories, double proteins, double fats, double carbs, double water,
            List<MealComponentInput> components)
        {
            try
            {
                var request = new UpdateMealRequest
                {
                    MealId = mealId,
                    Name = name,
                    Description = description,
                    Recipe = recipe,
                    ImageUrl = imageUrl,
                    Calories = calories,
                    Proteins = proteins,
                    Fats = fats,
                    Carbs = carbs,
                    Water = water
                };
                request.Components.AddRange(components);
                await _client.UpdateMealAsync(request);
                return true;
            }
            catch
            {
                return false;
            }
        }

        public async Task<bool> DeleteMealAsync(string mealId)
        {
            try
            {
                await _client.DeleteMealAsync(new DeleteMealRequest { MealId = mealId });
                return true;
            }
            catch
            {
                return false;
            }
        }

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;
            _channel?.Dispose();
            GC.SuppressFinalize(this);
        }
    }
}

File: android/app/src/main/java/enzio/android/eatfit/data/remote/GrpcMealService.kt
package enzio.android.eatfit.data.remote

import enzio.android.eatfit.proto.*
import io.grpc.ManagedChannel
import io.grpc.okhttp.OkHttpChannelBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.TimeUnit

class GrpcMealService(
    private val serverUrl: String = "192.168.0.10",
    private val port: Int = 50051
) {
    private val channel: ManagedChannel = OkHttpChannelBuilder
        .forAddress(serverUrl, port)
        .usePlaintext()
        .keepAliveTime(30, TimeUnit.SECONDS)
        .build()

    private val stub = MealServiceGrpc.newBlockingStub(channel)

    suspend fun listMeals(userId: String, sortBy: String = "created_at", sortOrder: String = "desc"): List<MealData> = withContext(Dispatchers.IO) {
        try {
            val request = ListMealsRequest.newBuilder()
                .setUserId(userId)
                .setSortBy(sortBy)
                .setSortOrder(sortOrder)
                .setPage(1)
                .setPageSize(100)
                .build()
            stub.listMeals(request).mealsList
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun searchMeals(userId: String, query: String, sortBy: String = "created_at", sortOrder: String = "desc"): List<MealData> = withContext(Dispatchers.IO) {
        try {
            val request = SearchMealsRequest.newBuilder()
                .setUserId(userId)
                .setQuery(query)
                .setSortBy(sortBy)
                .setSortOrder(sortOrder)
                .setPage(1)
                .setPageSize(100)
                .build()
            stub.searchMeals(request).mealsList
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getMeal(mealId: String): MealData? = withContext(Dispatchers.IO) {
        try {
            val request = GetMealRequest.newBuilder().setMealId(mealId).build()
            stub.getMeal(request).meal
        } catch (e: Exception) {
            null
        }
    }

    suspend fun createMeal(
        userId: String, name: String, description: String, recipe: String,
        imageUrl: String, calories: Double, proteins: Double, fats: Double,
        carbs: Double, water: Double, components: List<MealComponentInput>
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = CreateMealRequest.newBuilder()
                .setUserId(userId)
                .setName(name)
                .setDescription(description)
                .setRecipe(recipe)
                .setImageUrl(imageUrl)
                .setCalories(calories)
                .setProteins(proteins)
                .setFats(fats)
                .setCarbs(carbs)
                .setWater(water)
                .addAllComponents(components)
                .build()
            stub.createMeal(request).success
        } catch (e: Exception) {
            false
        }
    }

    suspend fun updateMeal(
        mealId: String, name: String, description: String, recipe: String,
        imageUrl: String, calories: Double, proteins: Double, fats: Double,
        carbs: Double, water: Double, components: List<MealComponentInput>
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = UpdateMealRequest.newBuilder()
                .setMealId(mealId)
                .setName(name)
                .setDescription(description)
                .setRecipe(recipe)
                .setImageUrl(imageUrl)
                .setCalories(calories)
                .setProteins(proteins)
                .setFats(fats)
                .setCarbs(carbs)
                .setWater(water)
                .addAllComponents(components)
                .build()
            stub.updateMeal(request).success
        } catch (e: Exception) {
            false
        }
    }

    suspend fun deleteMeal(mealId: String): Boolean = withContext(Dispatchers.IO) {
        try {
            stub.deleteMeal(DeleteMealRequest.newBuilder().setMealId(mealId).build()).success
        } catch (e: Exception) {
            false
        }
    }

    fun close() {
        channel.shutdown()
    }
}